File: modules/ballerina-core/src/main/java/org/wso2/ballerina/core/model/expressions/GreaterThanExpression.java
/*
*  Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
*
*  WSO2 Inc. licenses this file to you under the Apache License,
*  Version 2.0 (the "License"); you may not use this file except
*  in compliance with the License.
*  You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing,
*  software distributed under the License is distributed on an
*  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
*  KIND, either express or implied.  See the License for the
*  specific language governing permissions and limitations
*  under the License.
*/
package org.wso2.ballerina.core.model.expressions;

import org.wso2.ballerina.core.interpreter.Context;
import org.wso2.ballerina.core.model.NodeVisitor;
import org.wso2.ballerina.core.model.values.BValue;
import org.wso2.ballerina.core.model.values.BValueRef;
import org.wso2.ballerina.core.model.values.BooleanValue;
import org.wso2.ballerina.core.utils.TriFunction;

import static org.wso2.ballerina.core.model.Operator.GREATER_THAN;


/**
 * {@code GreaterThanExpression} represents a greater than (>) expression in Ballerina
 *
 * @since 1.0.0
 */
public class GreaterThanExpression extends BinaryCompareExpression {

    public static final TriFunction<Context, Expression, Expression, BValueRef> GREATER_THAN_INT_FUNC =
            (ctx, lExpr, rExpr) -> {
                boolean result = lExpr.evaluate(ctx).getInt() > rExpr.evaluate(ctx).getInt();
                BValue bValue = new BooleanValue(result);
                return new BValueRef(bValue);
            };

    public static final TriFunction<Context, Expression, Expression, BValueRef> GREATER_THAN_LONG_FUNC =
            (ctx, lExpr, rExpr) -> {
                boolean result = lExpr.evaluate(ctx).getLong() > rExpr.evaluate(ctx).getLong();
                BValue bValue = new BooleanValue(result);
                return new BValueRef(bValue);
            };

    public static final TriFunction<Context, Expression, Expression, BValueRef> GREATER_THAN_FLOAT_FUNC =
            (ctx, lExpr, rExpr) -> {
                boolean result = lExpr.evaluate(ctx).getFloat() > rExpr.evaluate(ctx).getFloat();
                BValue bValue = new BooleanValue(result);
                return new BValueRef(bValue);
            };

    public static final TriFunction<Context, Expression, Expression, BValueRef> GREATER_THAN_DOUBLE_FUNC =
            (ctx, lExpr, rExpr) -> {
                boolean result = lExpr.evaluate(ctx).getDouble() > rExpr.evaluate(ctx).getDouble();
                BValue bValue = new BooleanValue(result);
                return new BValueRef(bValue);
            };

    public GreaterThanExpression(Expression lExpr, Expression rExpr) {
        super(lExpr, GREATER_THAN, rExpr);
    }

    @Override
    public void accept(NodeVisitor visitor) {
        visitor.visit(this);
    }
}

File: modules/editor/web/js/ballerina/ast/greater-than-expression.js
/**
 * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
define(['lodash', './expression'], function (_, Expression) {

    /**
     * Constructor for GreaterThanExpression
     * @param {Object} args - Arguments to create the GreaterThanExpression
     * @constructor
     * @augments Expression
     */
    var GreaterThanExpression = function (args) {
        Expression.call(this, 'GreaterThanExpression');
    };

    GreaterThanExpression.prototype = Object.create(Expression.prototype);
    GreaterThanExpression.prototype.constructor = GreaterThanExpression;

    /**
     * setting parameters from json
     * @param {Object} jsonNode to initialize from
     */
    GreaterThanExpression.prototype.initFromJson = function (jsonNode) {
        this.setExpression(this.generateGreaterThanExpressionString(jsonNode));
    };

    /**
     * Generates the greater than expression as a string.
     * @param {Object} jsonNode - A node explaining the structure of greater than expression.
     * @return {string} - Arguments as a string.
     * @private
     */
    GreaterThanExpression.prototype.generateGreaterThanExpressionString = function (jsonNode) {
        var self = this;
        var greaterThanString = "";

        for (var itr = 0; itr < jsonNode.children.length; itr++) {
            var childJsonNode = jsonNode.children[itr];
            //TODO : Need to remove this if/else ladder by delegating expression string calculation to child classes
            if (childJsonNode.type == "basic_literal_expression") {
                if(childJsonNode.basic_literal_type == "string") {
                    // Adding double quotes if it is a string.
                    greaterThanString += "\"" + childJsonNode.basic_literal_value + "\"";
                } else {
                    greaterThanString += childJsonNode.basic_literal_value;
                }
            } else if (childJsonNode.type == "variable_reference_expression") {
                greaterThanString += childJsonNode.variable_reference_name;
            }

            if (itr !== jsonNode.children.length - 1) {
                greaterThanString += " < ";
            }
        }
        return greaterThanString;
    };

    return GreaterThanExpression;
});
